### ISSUE TITLE
Automatically install tweaker mods

### ISSUE THREAD

**Issue by Guichaguri** [MultiMC/MultiMC5#1777]
System Information
-----------------------------
MultiMC version: 0.4.11-751

Operating System: Windows 10

Summary of the issue or suggestion:
----------------------------------------------
Tweaker mods can provide a few entries in their manifest. Forge and LiteLoader read them and load the tweakers from the mods folder.
When Forge and LiteLoader are not present, tweaker mods can still be loaded by adding them as a library and adding the `--tweakClass` argument.
Currently, these tweakers can be easily installed in the vanilla launcher by bundled installers, but not in MultiMC.

What should happen:
------------------------------
MultiMC should check jar mods to register them as tweakers based on the information contained in the manifest.

Steps to reproduce the issue:
-------------------------------------------------------------
1. Add a tweaker (BetterFps or OptiFine) as a jar mod
2. Initialize the game
3. Nothing is loaded

Suspected cause:
---------------------------

Logs/Screenshots:
----------------------------
[//]: # (Please refer to https://github.com/MultiMC/MultiMC5/wiki/Log-Upload for instructions on how to attach your logs.)

Additional Info:
---------------------------
Here are the manifest entries that Forge and/or LiteLoader recognizes and a tweaker might have:
* `TweakClass` The class that implements the basic structure of a tweaker.
* `TweakOrder` The order which the tweaker should be loaded.
* `TweakName` The name of the mod.
* `TweakAuthor` The author of the mod.
* `TweakVersion` The version of the mod.

**Comment by peterix** [comment]
I do not see how this could work without pure guesswork about which dependencies those tweaker libraries have.

**Comment by peterix** [comment]
Example:
https://github.com/MultiMC/MultiMC5/wiki/MultiMC-and-OptiFine
I had to fix OptiFine because it was unable to even load as a tweaker properly.

This got fixed in later versions:
https://gist.github.com/peterix/5c8b655ffafb866ccb2519f421dab671

But it is not fixed in all versions.

**Comment by peterix** [comment]
Up to now, MultiMC has been using json files provided by tweaker authors to do this kind of thing. It is not ideal.

Forge - sadly - uses installers. The only reason why it is even feasible to install it is that the installers actually contain a barely workable json file with dependencies. It is missing actual download locations of the libraries. It does not mention that some of said libraries have to be first uncompressed and that the compression scheme is hacked up rather randomly.
Basically, the Forge installer is doing things that are impossible to predict based on the file. I am guessing and blindly altering the files based on hardcoded rules.

Liteloader provides enough information to construct a viable json file for actual releases - as long as said releases do not magically change on the server side...
Its snapshots do not follow a normal versioning scheme and make it impossible to update the json file.

**Comment by peterix** [comment]
Does your tweaker mod have useful metadata that would allow it to be installed?

I do not consider that manifest as enough. Which version of launchwrapper is needed? Which version of launchwrapper is needed for every tweaker mod on every Minecraft version? Does it need ASM? Does it need other libraries?

I do not know.

**Comment by peterix** [comment]
So, some examples.

Liteloader:
http://dl.liteloader.com/versions/versions.json

Forge:
* install_profile.json extracted from forge-1.11.2-13.20.0.2206-installer.jar: https://gist.github.com/peterix/6841dcf0491d188a4e4638fd4fba7c8e
* Compare with this adapted version MultiMC produces: https://gist.github.com/peterix/703f8d73ca07c2698f1bbcfa6e34e6a3

Looks like they added more undefined elements since the last time I looked at it.

**Comment by peterix** [comment]
I think none of this belongs in MultiMC - it should not be magically assembling these files based on assumptions made at some random point in time. Assumptions that may not hold in the future.

Let's just say that unlike the very much standard interface of dropping `.jar` files into the `mods` folder, tweakers are more or less complete chaos.

I would like to work towards removing this mess. Let's say we copy what Mojang does with their versions and put json files that describe versions **completely** on a server somewhere. It is a decent design.

**Comment by peterix** [comment]
As for the request - I am very much against confusing jar mods with tweakers. They are not the same. When installed manually, they should not be treated as the same. In case of Optifine, there is a large range of versions that are both jar mods and tweakers. If the logic that adds them makes assumptions - and makes it impossible to add certain versions of Optifine as real jar mods, I would consider that to a regression, not an improvement.

**Comment by peterix** [comment]
I'd like to know what you think about the above.

**Comment by Guichaguri** [comment]
I agree with you.
Even though no tweaker should require anything more than launchwrapper and ASM (all other dependencies should be included in the jar), you can't even provide the latest release of those libraries because newer versions of launchwrapper require joptsimple which don't even exist in older versions of the game (like beta). And I'm not even including Forge and LiteLoader, as they need special installation anyway.

So let me change this suggestion a bit: What if tweakers could provide a file (maybe multimc.json) in their jar files that included all the [necessary information](https://github.com/MultiMC/MultiMC5/wiki/MultiMC-and-OptiFine) for MultiMC? MultiMC could simply copy it instead of creating its own file

**Comment by peterix** [comment]
That could work. It would also be tightly coupled with what it is describing. That has its advantages and disadvantages.
One of the use cases of the Mojang design is that they update the versions with new versions of libraries. They replace the `realms` jar quite often in whatever Minecraft version is current.

But I'm fine with whatever you can provide that is accurate enough so that dumb automation can do the rest.

What I actually want to do in the near future is:
1. Stop putting the generated json files in instances by default. Instead, put them in a folder (or database) shared by all local instances.
2. Start moving these entirely out of MultiMC and replace the built in logic with something generic that accesses a repository of json version files.

What I'm concerned about is people who depend on users visiting their web pages for ad revenue.

@sp614x what would you prefer? The more recent versions of Optifine are much harder to install standalone now that they are not jar mods.

**Comment by sp614x** [comment]
@peterix do you mean OptiFine standalone in MultiMC?
Because it should still be loadable by Forge as a tweaker mod.

**Comment by peterix** [comment]
Yes. I mean standalone. This is all about standalone tweaker/jar mods, added on top of vanilla without any loaders.

With jar mods, it was easy. You either dragged it to a box (one click & drag), or browsed for a file (several mouse clicks). However, when it requires more actions, some people simply can't figure it out even with precise instructions.

**Comment by peterix** [comment]
Anyway more context and ideas...

For tweakers, the current stable version requires writing a version patch (json file in `$instanceRoot/patches/`). If all the files inside can be automatically downloaded, this is all there is to it. Still, the jsons don't have to be available on gist for all versions, people make typos, etc.
If a library cannot be easily downloaded, it needs to be marked as 'local' so MultiMC doesn't try to grab it from the Mojang repo. It also needs to be placed in `$dataRoot/libraries/long/package/path/name/version/name-version.extension`. You can imagine how well that goes.


------
The development version also searches for 'local' libraries inside instances ($instanceRoot/libraries/name-version.extension) - that's less annoying already.
It means you can grab the instance and move it to a different machine without messing with the global libraries, making it a lot more modpack-friendly.

It will have to be made obvious though, otherwise people won't be able to discover it. Maybe something along the lines of *'local library is missing -> show a dialog that prompts user to put the file in the libraries folder'*.

Issues: 
* Optifine has an underscore instead of a dash - it still requires a rename to be recognized.
* It does not fix the issue with people having to write json files.

----


The other idea I've been considering is recognizing files by hashes and grabbing the required metadata based on that from a server. Obviously, I would have to have a database with an index that maps file hashes back to the metadata they are associated with.
Examples:
* The shaders mod is added to the `mods` folder. It is recognized as the shaders mod using its sha256 hash. The metadata for it is downloaded and added as a version patch. The user gets a shaders page in the edit instance window.
* User adds optifine to the `libraries` folder. It is recognized as some version of optifine using its sha256 hash. The metadata for it is downloaded and added as a version patch. MultiMc can get its dependencies and it actually launches properly.

I'd have to build some infrastructure for this. Forge and Liteloader can be automated just as they are inside MultiMC now.

**Comment by peterix** [comment]
I've made some good progress on this. Current develop uses a metadata repository that  other things could be added to.

I will have to find a way to make installing packages that do not have direct downloads easy for the user. The current thought is that it will download files in the background and offer a window the other files can be dropped into. The files will be recognized by hash.

So, to complete the install, the user will be able to go to (for example) the OptiFine website, get the jar(s) using a browser and drop it on top of MultiMC.
